### PR TITLE
GPL-967: Cherrypicker generate data

### PR DESCRIPTION
**Pull request by sdjmchattie**
Adds work towards #322 

Changes proposed in this pull request:

* Add a POST endpoint for the cherrypicker generate data
* Stub out some responses but do some real analysis of the input parameters

I've added tests for the analysis helper method but not yet for the actual endpoint which I'll do when it's not just a stub

### CODE REVIEW

**Comment by sdjmchattie** [comment]
Right then @harrietc52 and @pjvv -- I think that's everything set up in Eve and fully tested

File: lighthouse/constants/fields.py [hunk 1]
```diff
@@ -1,109 +1,115 @@
 # Fields for baracoda
```
Comment (pjvv) [inline lighthouse/constants/fields.py:1]: Thanks for making this file better, could you sort alphabetically too?

File: lighthouse/helpers/responses.py [hunk 2]
```diff
@@ -4,11 +4,11 @@
 from lighthouse.types import FlaskResponse
 
 
-def bad_request(errors: Union[str, List[str]]) -> FlaskResponse:
+def bad_request(errors: Union[str, List[str]], **kwargs: Any) -> FlaskResponse:
```
Comment (pjvv) [inline lighthouse/helpers/responses.py:7]: Do we want to be returning anything other than a list of errors?

File: lighthouse/hooks/cherrypick_test_data.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+)
+from lighthouse.constants.error_messages import ERROR_CRAWLER_HTTP_ERROR
+from lighthouse.constants.fields import FIELD_CPTD_STATUS, FIELD_MONGO_ID
+
+logger = logging.getLogger(__name__)
+
+
+def insert_cherrypick_test_data_hook(runs: List[Dict[str, Any]]) -> None:
+    for run in runs:
+        run[FIELD_CPTD_STATUS] = CPTD_STATUS_PENDING
```
Comment (pjvv) [inline lighthouse/hooks/cherrypick_test_data.py:24]: Is it possible to set a default value in the schema instead or am I missing something?

File: lighthouse/hooks/cherrypick_test_data.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+
+def inserted_cherrypick_test_data_hook(runs: List[Dict[str, Any]]) -> None:
+    run_id = runs[0][FIELD_MONGO_ID]  # bulk inserting is disabled so there will only be one
```
Comment (pjvv) [inline lighthouse/hooks/cherrypick_test_data.py:28]: Maybe add to the comment where it is disabled?

File: lighthouse/hooks/cherrypick_test_data.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    try:
+        crawler_url = f"{app.config['CRAWLER_BASE_URL']}/cherrypick-test-data"
+        logger.debug(crawler_url)
```
Comment (pjvv) [inline lighthouse/hooks/cherrypick_test_data.py:32]: Seems like a forgotten log...

File: lighthouse/hooks/cherrypick_test_data.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        response = requests.post(crawler_url, json={FIELD_CRAWLER_RUN_ID: str(run_id)})
```
Comment (pjvv) [inline lighthouse/hooks/cherrypick_test_data.py:33]: Why convert the `run_id` to str?

File: lighthouse/types.py [hunk 7]
```diff
@@ -12,6 +12,11 @@
 EventMessage = Dict[str, Union[str, Event]]
 
 
+class EndpointParamsException(Exception):
```
Comment (pjvv) [inline lighthouse/types.py:15]: I don't think this is the right module for exceptions and I think the convention (not that it is always followed here) is to name the exception ending with `Error`, for instance `ValueError`, `KeyError`, `TypeError` etc.

File: tests/endpoints/cherrypick_test_data/test_valid_input.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+ENDPOINT_PATH = "/cherrypick-test-data"
+
+
+@pytest.fixture
+def mock_post():
+    with patch("requests.post") as mock_post:
+        yield mock_post
+
+
+def mock_response(status_code, json_object, http_error_msg=""):
```
Comment (pjvv) [inline tests/endpoints/cherrypick_test_data/test_valid_input.py:25]: I think you might find https://github.com/getsentry/responses helpful, we use it somewhere else too.

File: lighthouse/helpers/responses.py (hunk 2 above)
Comment (sdjmchattie) [reply lighthouse/helpers/responses.py:7]: Sorry, this was something I added when I was writing the API myself before using Eve.  I'll undo this change.

File: lighthouse/hooks/cherrypick_test_data.py (hunk 3 above)
Comment (sdjmchattie) [reply lighthouse/hooks/cherrypick_test_data.py:24]: I remember now that we discussed this.  There was a lot to take in at the time.  I've changed it.

File: lighthouse/hooks/cherrypick_test_data.py (hunk 5 above)
Comment (sdjmchattie) [reply lighthouse/hooks/cherrypick_test_data.py:32]: True ... I've now added more useful logging to this file.

File: lighthouse/hooks/cherrypick_test_data.py (hunk 6 above)
Comment (sdjmchattie) [reply lighthouse/hooks/cherrypick_test_data.py:33]: It's an ObjectId from Mongo.  You get an exception about it not being JSON serialisable if you don't turn it to a plain string.

File: lighthouse/types.py (hunk 7 above)
Comment (sdjmchattie) [reply lighthouse/types.py:15]: This was a left over from my first attempt.  I should have done a better job of stripping out the fluff that created!

File: tests/endpoints/cherrypick_test_data/test_valid_input.py (hunk 8 above)
Comment (sdjmchattie) [reply tests/endpoints/cherrypick_test_data/test_valid_input.py:25]: Why didn't I see this before?  I searched all over for something like this and didn't want to introduce any new dependencies to the project.  This is exactly what I needed and already here!

File: tests/endpoints/cherrypick_test_data/test_valid_input.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+import responses
+
+from lighthouse.constants.cherrypick_test_data import (
+    CPTD_STATUS_PENDING,
+    FIELD_CRAWLER_RUN_ID,
+)
+from lighthouse.constants.error_messages import ERROR_CRAWLER_HTTP_ERROR
+
+CRAWLER_URL = "http://127.0.0.1:8100/cherrypick-test-data"
+ENDPOINT_PATH = "/cherrypick-test-data"
```
Comment (pjvv) [inline tests/endpoints/cherrypick_test_data/test_valid_input.py:17]: Seems like some duplication, you could move `ENDPOINT_PATH` before and create `CRAWLER_URL` using it.

**Review by pjvv — APPROVED** [review]

File: tests/endpoints/cherrypick_test_data/test_valid_input.py (hunk 9 above)
Comment (sdjmchattie) [reply tests/endpoints/cherrypick_test_data/test_valid_input.py:17]: They refer to different endpoints.  The Crawler one is an external endpoint.  The second one is the endpoint in this app under test.

File: lighthouse/types.py (hunk 7 above)
Comment (pjvv) [reply lighthouse/types.py]: Haha no worries.

File: tests/endpoints/cherrypick_test_data/test_valid_input.py (hunk 8 above)
Comment (pjvv) [reply tests/endpoints/cherrypick_test_data/test_valid_input.py]: Cool, I actually forgot it was already in the project too.

File: tests/endpoints/cherrypick_test_data/test_valid_input.py (hunk 9 above)
Comment (pjvv) [reply tests/endpoints/cherrypick_test_data/test_valid_input.py:17]: OK, so you are keeping them separate for clarity sake?

Comment (sdjmchattie) [reply tests/endpoints/cherrypick_test_data/test_valid_input.py:17]: I don't think they should be merged.  The endpoints only happen to have the same path.  It's not a guaranteed future certainty.

Comment (pjvv) [reply tests/endpoints/cherrypick_test_data/test_valid_input.py:17]: Cool cool.

**Pull request merged**